src/main/java/ch/njol/skript/lang/function/FunctionRegistry.java

```diff
@@ -0,0 +1,701 @@
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.SkriptAPIException;
+import com.google.common.base.Preconditions;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnmodifiableView;
+import org.skriptlang.skript.lang.converter.Converters;
+import org.skriptlang.skript.util.Registry;
+
+import java.util.*;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * A registry for functions.
+ */
+final class FunctionRegistry implements Registry<Function<?>> {
+
+	private static FunctionRegistry registry;
+
+	/**
+	 * Gets the global function registry.
+	 *
+	 * @return The global function registry.
+	 */
+	public static FunctionRegistry getRegistry() {
+		if (registry == null) {
+			registry = new FunctionRegistry();
+		}
+		return registry;
+	}
+
+	/**
+	 * The pattern for a valid function name.
+	 * Functions must start with a letter or underscore and can only contain letters, numbers, and underscores.
+	 */
+	final static String FUNCTION_NAME_PATTERN = "[\\p{IsAlphabetic}_][\\p{IsAlphabetic}\\d_]*";
+
+	/**
+	 * The namespace for functions registered using Java.
+	 */
+	private final NamespaceIdentifier GLOBAL_NAMESPACE = new NamespaceIdentifier(Scope.GLOBAL, null);
+
+	/**
+	 * All registered namespaces.
+	 */
+	private final Map<NamespaceIdentifier, Namespace> namespaces = new ConcurrentHashMap<>();
+
+	@Override
+	@UnmodifiableView
+	public @NotNull Collection<Function<?>> elements() {
+		Set<Function<?>> functions = new HashSet<>();
+
+		for (Namespace namespace : namespaces.values()) {
+			functions.addAll(namespace.functions.values());
+		}
+
+		return Collections.unmodifiableSet(functions);
+	}
+
+	/**
+	 * Registers a signature.
+	 *
+	 * @param namespace The namespace to register the signature in.
+	 *                  If namespace is null, will register this signature globally.
+	 *                  Usually represents the path of the script this signature is registered in.
+	 * @param signature The signature to register.
+	 * @throws SkriptAPIException if a signature with the same name and parameters is already registered
+	 *                            in this namespace.
+	 */
+	public void register(@Nullable String namespace, @NotNull Signature<?> signature) {
+		Preconditions.checkNotNull(signature, "signature cannot be null");
+		Skript.debug("Registering signature '%s'", signature.getName());
+
+		// namespace
+		NamespaceIdentifier namespaceId;
+		if (namespace != null && signature.isLocal()) {
+			namespaceId = new NamespaceIdentifier(Scope.LOCAL, namespace);
+		} else {
+			namespaceId = GLOBAL_NAMESPACE;
+		}
+
+		// since we are getting a namespace and then updating it by putting it back in the map,
+		// avoid race conditions by ensuring only one thread can access the namespaces map at a time.
+		synchronized (namespaces) {
+			namespaces.compute(namespaceId, (id, ns) -> {
```

Comment:
Rather than putting all of this in a compute call, you could just `computeIfAbsent` to create the namespace, right?